src/tribler-core/tribler_core/modules/metadata_store/community/gigachannel_community.py

```diff
@@ -117,11 +164,32 @@ def notify_gui(_, processing_results):
             if self.notifier and results:
                 self.notifier.notify(NTFY.REMOTE_QUERY_RESULTS, {"results": results, "uuid": str(request_uuid)})
 
-        for p in self.get_random_peers(self.settings.max_query_peers):
+        # Try sending the request to at least some peers that we know have it
+        if "channel_pk" in kwargs and "origin_id" in kwargs:
+            peers_to_query = self.get_known_subscribed_peers_for_node(
+                unhexlify(kwargs["channel_pk"]), kwargs["origin_id"], self.settings.max_mapped_query_peers
+            )
+        else:
+            peers_to_query = self.get_random_peers(self.settings.max_query_peers)
+
+        for p in peers_to_query:
             self.send_remote_select(p, **kwargs, processing_callback=notify_gui)
 
         return request_uuid
 
+    def get_known_subscribed_peers_for_node(self, node_pk, node_id, limit=None):
```

Comment:
**improvement:** seems that you go to the database for each request.

What do you think about caching this information to prevent frequent access to the database?

Reply:
Caching these results can break the logic in some cases, such as when the channel author moves some intermediary folders to another channel. Or when our instance fails to find the root node correctly the first time (e.g. after a search), but then receives the missing ancestor nodes. Unless we build a subsystem for caching the stuff into the MetadataStore itself, along with a way to **invalidate** the caching results, but that's worth a different PR, with an extensive test suite. Another way to solve it would be to add transitive closures to node entries.

That said, it should be harmless in 99.99% of cases to cache the thing. My only real concern is how to indicate to the programmers (basically, *us*) in the future, that the thing is unreliable in some corner cases? Argument flags? Big warning signs?

I would like to hear your suggestions.